samples/cpp/digits.cpp

```diff
@@ -0,0 +1,468 @@
+    "their accuracy.\n"
+    "\nFollowing preprocessing is applied to the dataset:\n"
+    " - Moment-based image deskew (see deskew())\n"
+    " - Digit images are split into 4 10x10 cells and 16-bin\n"
+    "   histogram of oriented gradients is computed for each\n"
+    "   cell\n"
+    " - Transform histograms to space with Hellinger metric (see [1] (RootSIFT))\n"
+    "\n\n[1] R. Arandjelovic, A. Zisserman\n"
+    "    \"Three things everyone should know to improve object retrieval\"\n"
+    "    http://www.robots.ox.ac.uk/~vgg/publications/2012/Arandjelovic12/arandjelovic12.pdf\n"
+    "\nUsage:\n"
+    "   ./digits\n" << endl;
+}
+
+static void split2d(Mat& image, Size cell_size, vector<Mat>& cells)
+{
+    int height = image.rows;
+    int width = image.cols;
+
+    int sx = cell_size.width;
+    int sy = cell_size.height;
+
+    cells.clear();
+
+    for (int i = 0; i < height; i += sy)
+    {
+        for (int j = 0; j < width; j += sx)
+        {
+            cells.push_back(image(Rect(j, i, sx, sy)).clone());
+        }
+    }
+}
+
+static void load_digits(const char* fn, vector<Mat>& digits, vector<int>& labels)
+{
+    digits.clear();
+    labels.clear();
+
+    String filename = samples::findFile(fn);
+
+    cout << "Loading " << filename << " ..." << endl;
+
+    Mat digits_img = imread(filename, IMREAD_GRAYSCALE);
+    split2d(digits_img, Size(SZ, SZ), digits);
+
+    for (int i = 0; i < CLASS_N; i++)
+    {
+        for (size_t j = 0; j < digits.size() / CLASS_N; j++)
+        {
+            labels.push_back(i);
+        }
+    }
+}
+
+static void deskew(Mat& img, Mat& deskewed_img)
+{
+    Moments m = moments(img);
+
+    if (abs(m.mu02) < 0.01)
+    {
+        deskewed_img = img.clone();
+        return;
+    }
+
+    float skew = (float)(m.mu11 / m.mu02);
+    float M_vals[2][3] = {{1, skew, -0.5f * SZ * skew}, {0, 1, 0}};
+    Mat M(Size(3, 2), CV_32F);
+
+    for (int i = 0; i < M.rows; i++)
+    {
+        for (int j = 0; j < M.cols; j++)
+        {
+            M.at<float>(i, j) = M_vals[i][j];
+        }
+    }
+
+    warpAffine(img, deskewed_img, M, Size(SZ, SZ), WARP_INVERSE_MAP | INTER_LINEAR);
+}
+
+class StatModel
+{
+    public:
+        virtual void load(const char* fn) = 0;
+
+        virtual void save(const char* fn) = 0;
+
+        virtual void train(Mat& samples, vector<int>& responses) = 0;
+
+        virtual void predict(Mat& samples, vector<float>& results) = 0;
+
+        virtual ~StatModel() {}
+};
+
+class KNearest : public StatModel
+{
+    private:
+        int _k;
+        Ptr<ml::KNearest> _model;
+
+    public:
+        KNearest(int k = 3)
+        {
+            _k = k;
+            _model = ml::KNearest::create();
+        }
+
+        static Ptr<KNearest> create(int k = 3)
+        {
+            return makePtr<KNearest>(k);
+        }
+
+        void load(const char* fn)
+        {
+            _model->load<ml::KNearest>(fn);
```

Comment:
this won't work (same for SVM).
`Algorithm::load()` is a static method, and returns a new instance, 

[see here for a usage example](https://github.com/opencv/opencv/blob/master/samples/cpp/letter_recog.cpp#L88)